src/hotspot/share/prims/forte.hpp

```diff
@@ -34,4 +36,22 @@ class Forte : AllStatic {
                                                  // register internal VM stub
 };
 
+// A small RAII mark class to manage 'in_asgct' thread status
+class ASGCTMark : public StackObj {
+ private:
+  // NONCOPYABLE(ASGCTMark);
+  ASGCTMark(JavaThread* thread) {
+    if (thread != nullptr) {
+      assert(thread == JavaThread::current(), "not the current thread");
+      thread->set_in_asgct(true);
+    }
+  }
+
+ public:
+  ASGCTMark() : ASGCTMark(JavaThread::current()) {}
+  ~ASGCTMark() {
+    JavaThread::current()->set_in_asgct(false);
```

Comment:
You can't call `JavaThread::current()` in any of this code as it is not safe from a signal handling context. (There is an existing use in ASGCT that is also unsafe.) I suggest not having the no-arg constructor and require the current JavaThread, or null, to be passed in (having already been safely obtained by the caller). You can then assert using `Thread::current_or_null_safe()`.

Personally I find the ASGCTMark complete overkill for this situation as there is only ever going to be a single use - sorry @tstuefe - it is just complicating things IMO.

Reply:
Hi David,

> You can't call `JavaThread::current()` in any of this code as it is not safe from a signal handling context. (There is an existing use in ASGCT that is also unsafe.) I suggest not having the no-arg constructor and require the current JavaThread, or null, to be passed in (having already been safely obtained by the caller). You can then assert using `Thread::current_or_null_safe()`.
> 
> Personally I find the ASGCTMark complete overkill for this situation as there is only ever going to be a single use - sorry @tstuefe - it is just complicating things IMO.

No problem, that's why we talk :)

To make an argument for this: if you compare the first iterations of this patch to the current one, the current one is a lot simpler. I really dislike funneling arguments through many layers down to a utility function, only to fine control one specific behavioral aspect of that tiny function, and that aspect is only relevant for one outlier use case. I dislike this more than the RAII object.

Look at it this way. We have a thread-specific state (we are in AGCT and deep down somewhere we want to do something different depending on that state). We need to pass that information down. The only difference is where we keep the "we are in thread-specific state" info - on the stack as an argument chain, or anchored to TLS.

Putting it into arguments is one way, but I find this rather ugly. Now every caller of this function has to care. With the RAII, we have the information as a mark exactly where it matters - in the entrance frame of AGCT - and no outside code needs to care.

Cheers, Thomas

Reply:
But you aren't keeping it on the stack versus TLS you are using a stack object to set the TLS. All we have done in this final form is replace:

```
thread->set_in_asgct(true);
<stuff>
thread->set_in_asgct(false);
```
with

```
ASGCTMark mark(thread);
<stuff>
```
but now ASGCTMark has to worry about whether `thread` is NULL, whether it is the current javaThread, when we've already done all that for `thread`.

And this is the only place we will ever use ASGCTMark. Now if we had lots of return points in `<stuff>` then RAII would definitely be better to ensure we can't forget to reset the state. But we don't. So use of RAII adds complexity cost now just in case in the future RAII might be useful.

Reply:
Fair enough.

I removed the RAII and added a comment making explicit the requirement to reset the 'in_asgct' flag on method return. If there ever appear more return points in that method the RAII can be reconsidered.

Reply:
It would be interesting to see if it this ASGCT flag is useful in JFR, i.e. if JFR suffers potentially from the same problem. There would then be two additional places for setting this flag.

Reply:
AFAIK, currently JFR is 'wrapping' the code in a crash-handler.
We can take a look at reusing this approach there in some follow-up PR but for now I would really prefer getting this one merged without attaching any more bells&whistles.

Reply:
But the crash-handler does not prevent that `guarantee` aborts the VM, as we discussed already in my related PR. My comment was more a suggestion and related to "And this is the only place we will ever use ASGCTMark" by @dholmes-ora . I agree with you that changing JFR should be done in another PR.

Reply:
I thought that "this approach" referred to the approach of this PR (possibly renaming the flag), as CrashProtection is orthogonal to the issue that this PR solves.

Reply:
> thought that "this approach" referred to the approach of this PR (possibly renaming the flag)
^ this

But still, I would prefer to defer any further improvements into a separate PR.